### PR TITLE
Aave V3 Ethereum: Add CRV (draft)

### PR DESCRIPTION
**Pull request by efecarranza**
**Pending Llama internal review** 

# Changelog

Payload to add CRV to Aave v3 Ethereum.
Tests files.
Diff files.

### CODE REVIEW

File: src/contracts/mainnet/AaveV3EthAddCRVPoolPayload.sol [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+  constructor() AaveV3ListingEthereum(IGenericV3ListingEngine(AaveV3Ethereum.LISTING_ENGINE)) {}
+
+  function getAllConfigs() public pure override returns (IGenericV3ListingEngine.Listing[] memory) {
+    IGenericV3ListingEngine.Listing[] memory listings = new IGenericV3ListingEngine.Listing[](1);
+
+    listings[0] = IGenericV3ListingEngine.Listing({
+      asset: AaveV2EthereumAssets.CRV_UNDERLYING,
+      assetSymbol: 'CRV',
+      priceFeed: AaveV2EthereumAssets.CRV_ORACLE,
```
Comment (0xrajath) [inline src/contracts/mainnet/AaveV3EthAddCRVPoolPayload.sol:27]: I know this won't change. But just double checking if you should be using the AaveV2EthereumAssets here

File: src/contracts/mainnet/AaveV3EthAddCRVPoolPayload.sol [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+/**
+ * @title List CRV on AaveV3Ethereum
+ * @author Llama
+ * @dev This proposal lists CRV on Aave V3 Ethereum
+ * Governance: https://governance.aave.com/t/arfc-add-crv-to-ethereum-v3/11532
+ * Snapshot: https://snapshot.org/#/aave.eth/proposal/0xa0b5336692735fb4288537646aec5819cb4cbf01fc3a8a7cb06c9e62db708055
+ */
+contract AaveV3EthAddCRVPoolPayload is AaveV3ListingEthereum {
+  address public constant INTEREST_RATE_STRATEGY = 0x76884cAFeCf1f7d4146DA6C4053B18B76bf6ED14;
```
Comment (0xrajath) [inline src/contracts/mainnet/AaveV3EthAddCRVPoolPayload.sol:17]: I saw some term called `Base Stable Rate Offset` in https://snapshot.org/#/aave.eth/proposal/0xa0b5336692735fb4288537646aec5819cb4cbf01fc3a8a7cb06c9e62db708055 which doesn't seem to be in this interest rate strategy. 

Is that as intended?

File: src/test/mainnet/AaveV3EthAddCRVPoolPayloadTest.t.sol [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+import 'forge-std/Test.sol';
+import {AaveV2EthereumAssets} from 'aave-address-book/AaveV2Ethereum.sol';
+import {AaveV3Ethereum, AaveV3EthereumAssets} from 'aave-address-book/AaveV3Ethereum.sol';
+import {ProtocolV3TestBase, InterestStrategyValues, ReserveConfig} from 'aave-helpers/ProtocolV3TestBase.sol';
+import {AaveGovernanceV2} from 'aave-address-book/AaveGovernanceV2.sol';
+import {TestWithExecutor} from 'aave-helpers/GovHelpers.sol';
+import {AaveV3EthAddCRVPoolPayload} from '../../contracts/mainnet/AaveV3EthAddCRVPoolPayload.sol';
+
+contract AaveV3EthAddCRVPoolPayloadTest is ProtocolV3TestBase, TestWithExecutor {
```
Comment (0xrajath) [inline src/test/mainnet/AaveV3EthAddCRVPoolPayloadTest.t.sol:13]: Should we have some integration tests checking borrowing, depositing, etc. post-listing?

**Review by 0xrajath — APPROVED** [review]
LGTM. Left some comments

File: src/contracts/mainnet/AaveV3EthAddCRVPoolPayload.sol [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+  constructor() AaveV3ListingEthereum(IGenericV3ListingEngine(AaveV3Ethereum.LISTING_ENGINE)) {}
+
+  function getAllConfigs() public pure override returns (IGenericV3ListingEngine.Listing[] memory) {
+    IGenericV3ListingEngine.Listing[] memory listings = new IGenericV3ListingEngine.Listing[](1);
+
+    listings[0] = IGenericV3ListingEngine.Listing({
+      asset: AaveV2EthereumAssets.CRV_UNDERLYING,
```
[the remark this answers is not in the document]
Comment (efecarranza) [reply src/contracts/mainnet/AaveV3EthAddCRVPoolPayload.sol:25]: It doesn't exist in V3 yet to add so I think it makes sense to use the library where we know the values are right.

File: src/contracts/mainnet/AaveV3EthAddCRVPoolPayload.sol (hunk 1 above)
Comment (efecarranza) [reply src/contracts/mainnet/AaveV3EthAddCRVPoolPayload.sol:27]: It doesn't exist in V3 yet to add so I think it makes sense to use the library where we know the values are right.

File: src/contracts/mainnet/AaveV3EthAddCRVPoolPayload.sol [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+      assetSymbol: 'CRV',
+      priceFeed: AaveV2EthereumAssets.CRV_ORACLE,
+      rateStrategy: INTEREST_RATE_STRATEGY,
+      enabledToBorrow: true,
+      stableRateModeEnabled: false,
+      borrowableInIsolation: true,
+      withSiloedBorrowing: false,
```
[the remark this answers is not in the document]
Comment (efecarranza) [reply src/contracts/mainnet/AaveV3EthAddCRVPoolPayload.sol:32]: You're right, `borrowableInIsolation` should be false as well. 
Debt ceiling == isolation mode, not this parameter.

@sakulstra wanted to see if you could emphasize this area when reviewing to make sure I understood after your explanation last week. thank you.

File: src/contracts/mainnet/AaveV3EthAddCRVPoolPayload.sol (hunk 2 above)
Comment (efecarranza) [reply src/contracts/mainnet/AaveV3EthAddCRVPoolPayload.sol:17]: It's a bit tricky, as it's added to the base stable borrow rate. You can see it at the bottom of this page: https://etherscan.io/address/0x76884cAFeCf1f7d4146DA6C4053B18B76bf6ED14#code
under constructor arguments.

```
Arg [6] : stableRateSlope2 (uint256): 3000000000000000000000000000
```

File: src/test/mainnet/AaveV3EthAddCRVPoolPayloadTest.t.sol (hunk 3 above)
Comment (efecarranza) [reply src/test/mainnet/AaveV3EthAddCRVPoolPayloadTest.t.sol:13]: @sakulstra had mentioned last week that they don't think that's necessary for new listings as the protocol itself has been heavily tested.

File: src/contracts/mainnet/AaveV3EthAddCRVPoolPayload.sol (hunk 4 above)
[the remark this answers is not in the document]
Comment (0xrajath) [reply src/contracts/mainnet/AaveV3EthAddCRVPoolPayload.sol:25]: SGTM

File: src/contracts/mainnet/AaveV3EthAddCRVPoolPayload.sol (hunk 1 above)
Comment (0xrajath) [reply src/contracts/mainnet/AaveV3EthAddCRVPoolPayload.sol:27]: SGTM

**Review by 0xrajath — APPROVED** [review]
LGTM

File: src/test/mainnet/AaveV3EthAddCRVPoolPayloadTest.t.sol (hunk 3 above)
Comment (0xrajath) [reply src/test/mainnet/AaveV3EthAddCRVPoolPayloadTest.t.sol:13]: SGTM

File: src/contracts/mainnet/AaveV3EthAddCRVPoolPayload.sol (hunk 1 above)
Comment (efecarranza) [reply src/contracts/mainnet/AaveV3EthAddCRVPoolPayload.sol:27]: Leaving a question here for @sakulstra as he just pointed out on Telegram that the V2 Oracle is CRV/ETH and it being 10PM I'd just rather you see this tomorrow morning:

The AIP/spec didn't list the oracle so I wrongly assumed would be the same one as the V2 one. I now know I could have used the tenderly cli to see the UI and validate this but, if there's only an ETH based oracle, what other changes/set up needs to be done for this to work correctly on V3? 

Thank you.

File: src/contracts/mainnet/AaveV3EthAddCRVPoolPayload.sol [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    listings[0] = IGenericV3ListingEngine.Listing({
+      asset: AaveV2EthereumAssets.CRV_UNDERLYING,
+      assetSymbol: 'CRV',
+      priceFeed: CRV_USD_FEED,
+      rateStrategy: INTEREST_RATE_STRATEGY,
+      enabledToBorrow: true,
+      stableRateModeEnabled: false,
+      borrowableInIsolation: false,
+      withSiloedBorrowing: false,
+      flashloanable: false,
```
Comment (sakulstra) [inline src/contracts/mainnet/AaveV3EthAddCRVPoolPayload.sol:34]: should be true on 3.0.1 (so ethereum pool)

File: src/contracts/mainnet/AaveV3EthAddCRVPoolPayload.sol [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+      ltv: 55_00,
+      liqThreshold: 61_00,
+      liqBonus: 8_30,
+      reserveFactor: 20_00,
+      supplyCap: 62_500_000,
+      borrowCap: 7_700_000,
+      debtCeiling: 20_900_000,
```
Comment (sakulstra) [inline src/contracts/mainnet/AaveV3EthAddCRVPoolPayload.sol:41]: ceiling is wrong as it's with 2 decimals (what you wrote here means, 209k, not 20.9M) - you can check with tenderly cli

File: src/contracts/mainnet/AaveV3EthAddCRVPoolPayload.sol (hunk 1 above)
Comment (sakulstra) [reply src/contracts/mainnet/AaveV3EthAddCRVPoolPayload.sol]: it's not really a v3 vs v2 think with the oracle.

Aave can have different base units, but they are for the whole pool - aave v2 ethereum afaik is the only pool using eth base, every other v2 and v3 pool uses USD.

**Pull request merged**